client-app/src/desktop/tabs/other/routing/AdvancedRoutingPanel.tsx

```diff
@@ -0,0 +1,198 @@
+                            {value: 'company|desc', label: 'Company Name (Desc)'},
+                            {value: 'company|asc', label: 'Company Name (Asc)'},
+                            {value: 'city|desc', label: 'City (Desc)'},
+                            {value: 'city|asc', label: 'City (Asc)'},
+                            {value: 'trade_date|desc', label: 'Trade Date (Desc)'},
+                            {value: 'trade_date|asc', label: 'Trade Date (Asc)'},
+                            {value: null, label: 'None'}
+                        ]
+                    }),
+                    switchInput({
+                        label: 'Prevent Route Deactivation',
+                        onChange: () => (model.preventDeactivate = !model.preventDeactivate)
+                    }),
+                    filler(),
+                    vbox({})
+                ]
+            })
+        });
+    }
+});
+
+class AdvancedRoutingPanelModel extends HoistModel {
+    @observable groupBy = null;
+    @observable sortBy = null;
+    @observable preventDeactivate = false;
+
+    constructor() {
+        super();
+        makeObservable(this);
+
+        this.addReaction({
+            track: () => XH.routerState.params,
+            run: () => this.parseRouteParams(),
+            fireImmediately: true
+        });
+
+        this.addReaction({
+            track: () => [this.groupBy, this.sortBy, this.gridModel.selectedRecord?.id],
+            run: () => this.updateRoute(),
+            fireImmediately: true
+        });
+
+        window.addEventListener('beforeunload', e => {
+            if (!XH.routerState.name.startsWith('default.other.advancedRouting')) {
+                delete e.returnValue;
+                return;
+            }
+            if (this.preventDeactivate) e.preventDefault();
+        });
+    }
+
+    gridModel = new GridModel({
+        columns: [
+            {field: 'id', flex: 0},
+            {field: 'company', flex: 1},
+            {field: 'city', flex: 1},
+            {field: 'trade_date', flex: 1}
+        ]
+    });
+
+    @action
+    private setGroupBy(groupBy: string) {
+        this.groupBy = groupBy;
+
+        // Always select first when regrouping.
+        const groupByArr = groupBy ? groupBy.split(',') : [];
+        this.gridModel.setGroupBy(groupByArr);
+    }
+
+    @action
+    private setSortBy(sortBy: string) {
+        this.sortBy = sortBy;
+
+        // Always select first when resorting.
+        const sortByArr = sortBy ? sortBy.split(',') : [];
+        this.gridModel.setSortBy(sortByArr);
+    }
+
+    @action
+    private async setSelected(recordId: string | number) {
+        await this.gridModel.selectAsync(Number(recordId));
+        if (!this.gridModel.selectedId) {
+            XH.dangerToast(`Record ${recordId} not found`);
+        }
+    }
+
+    @action
+    private async parseRouteParams() {
+        if (this.gridModel.empty) return;
```

Comment:
Necessary?

Reply:
This is necessary because `parseRouteParams` is ran when the page initializes, before the grid model loads in its data. When removed, this will cause the state passed in via `routeParam` to be reset to the no-selection case.